lib/dpl/provider/pages.rb

```diff
@@ -44,34 +53,92 @@ def slug
         options.fetch(:repo) { context.env['TRAVIS_REPO_SLUG'] }
       end
 
+      def keep_history
+        options.fetch(:keep_history, false)
+      end
+
+      def allow_empty_commit
+        options.fetch(:allow_empty_commit, false)
+      end
+
+      def verbose
+        # Achtung! Never verbosify git, since it may expose user's token.
+        options.fetch(:verbose, false)
+      end
+
       def check_auth
       end
 
       def needs_key?
         false
       end
 
-      def github_deploy
-        context.shell 'rm -rf .git > /dev/null 2>&1'
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
-        context.shell 'git init' or raise 'Could not create new git repo'
+      def print_step(msg)
+        log msg if @verbose
+      end
+
+      def github_pull_or_init(target_dir)
+        unless @keep_history
+          github_init
+          return
+        end
+
+        print_step "Trying to clone a single branch #{@target_branch} from existing repo..."
+        unless context.shell "git clone --quiet --branch='#{@target_branch}' --depth=1 '#{@gh_remote_url}' '#{target_dir}' &>/dev/null"
+          # if such branch doesn't exist at remote, init it from scratch
+          print_step "Cloning #{@target_branch} branch failed"
+          Dir.mkdir(target_dir)  # Restore dir destroyed by failed `git clone`
+          github_init
+        end
+      end
+
+      def github_init
+        print_step 'Creating a brand new local repo from scratch...'
+        context.shell "git init" or raise 'Could not create new git repo'
+        print_step 'Repo created successfully'
+        context.shell "git checkout --orphan '#{@target_branch}'" or raise 'Could not create an orphan git branch'
+        print_step "An orphan branch #{@target_branch} created successfully"
+      end
+
+      def github_configure
+        print_step "Configuring git committer to be #{@gh_name} <#{@gh_email}>"
         context.shell "git config user.email '#{@gh_email}'"
         context.shell "git config user.name '#{@gh_name}'"
+      end
+
+      def github_commit
+        print_step "Preparing to deploy #{@target_branch} branch to gh-pages"
+        context.shell "touch \"deployed at `date` by #{@gh_name}\""
         context.shell "echo '#{@gh_fqdn}' > CNAME" if @gh_fqdn
-        context.shell 'git add .'
-        context.shell "FILES=\"`git commit -m 'Deploy #{@project_name} to #{@gh_ref}:#{@target_branch}' | tail`\"; echo \"$FILES\"; echo \"$FILES\" | [ `wc -l` -lt 10 ] || echo '...'"
-        context.shell "git push --force --quiet 'https://#{@gh_token}@#{@gh_ref}' master:#{@target_branch} > /dev/null 2>&1"
+        context.shell 'git add -A .'
```

Comment:
I believe you don't need the `-A` anymore in order to add removals to the index.

Reply:
@jyasskin the defaults for Git 1.x and 2.x are different.
And taking into account that dpl is also an open-source tool which can be freely run outside of @travis-ci infra (in the env with Git 1.x there), I would insist on keeping `-A`. [Explicit is better than implicit.](https://www.python.org/dev/peps/pep-0020/#the-zen-of-python)

Ref: https://stackoverflow.com/a/26039014/595220

Reply:
Ah, `-A` sounds good to me then. Thanks for explaining.